### PR TITLE
Thuang 10 type state manager

### PR DESCRIPTION
**Pull request by tihuan**
This PR covers `client/src/util/stateManager` with commits in diff! PTAL thank you 🙏

### CODE REVIEW

File: client/__tests__/util/stateManager/colorHelpers.test.ts
Comment (seve) [inline client/__tests__/util/stateManager/colorHelpers.test.ts:118]: why not:
```suggestion
        ct?.scale?.(cats.indexOf(data[i]))
```

File: client/src/actions/index.ts
Comment (seve) [inline client/src/actions/index.ts:39]: call `fetchJson` with a type to fulfill the type generic, this will guarantee the return type instead of casting

File: client/src/reducers/genesets.ts
Comment (seve) [inline client/src/reducers/genesets.ts:40]: export as less generic `GenesetsState`

Comment (seve) [reply client/src/reducers/genesets.ts:40]: Also type action

File: client/src/util/stateManager/colorHelpers.ts
Comment (seve) [inline client/src/util/stateManager/colorHelpers.ts:28]: Why `Action["type"]`? Isn't this just `any`

File: client/src/util/stateManager/colorHelpers.ts
Comment (seve) [inline client/src/util/stateManager/colorHelpers.ts:247]: What case were we seeing `createRgbArray` called with a single arg?

File: client/src/util/stateManager/controlsHelpers.ts
Comment (seve) [inline client/src/util/stateManager/controlsHelpers.ts:112]: Based off of [this comment describing CategoricalSelection](https://github.com/chanzuckerberg/single-cell-explorer/blob/8d21a5553ef33f37224b256fcf3f094efac8d8d7/client/src/reducers/categoricalSelection.ts#L3-L13), the Map can be typed so:
```suggestion
  [key: string]: Map<string, boolean>;
```

**Review by seve — CHANGES_REQUESTED** [review]
A few questions/comments

File: client/__tests__/util/stateManager/colorHelpers.test.ts
Comment (tihuan) [reply client/__tests__/util/stateManager/colorHelpers.test.ts:118]: Thanks for flagging! `scale` is a function, so we need to check the function exists before invoking it!

Unfortunately we can't do `ct?.scale?(cats.indexOf(data[i]))`, because the compiler thinks it's ternary operation or something -.-

File: client/src/actions/index.ts
Comment (tihuan) [reply client/src/actions/index.ts:39]: Ahh great idea! Thank you!

File: client/src/reducers/genesets.ts
Comment (tihuan) [reply client/src/reducers/genesets.ts:40]: Ah I didn't export State because no one is importing it at the moment, maybe we can keep it not exported until a callsite actually needs it?

Just added `AnyAction` for `action` 🙆‍♂️

File: client/src/util/stateManager/colorHelpers.ts
Comment (tihuan) [reply client/src/util/stateManager/colorHelpers.ts:28]: Technically `Action["type"]` is generic and can be strings or symbols. I opted to use `Action["type"]` here instead of `string` because I wanted to signal that `colorMode` comes from action type directly, which `string` can't convey such relationship. What do you think?

File: client/src/util/stateManager/colorHelpers.ts
Comment (tihuan) [inline client/src/util/stateManager/colorHelpers.ts:227]: this `colors` can be `undefined`

File: client/src/util/stateManager/colorHelpers.ts
Comment (tihuan) [reply client/src/util/stateManager/colorHelpers.ts:247]: Great question! `colors` is optional here because callsite could be undefined as `colors`. E.g., https://github.com/chanzuckerberg/single-cell-explorer/pull/19/files/bfaa6ac79d14925936517da37631bc6700bf5538#r705659559

File: client/src/util/stateManager/controlsHelpers.ts
Comment (tihuan) [reply client/src/util/stateManager/controlsHelpers.ts:112]: Oh thank you for linking that! Updated 🙆‍♂️

**Comment by tihuan** [comment]
Just updated the PR! Thanks so much for all the helpful feedback, @seve 🙏 🙏 🙏 !

File: client/__tests__/util/stateManager/colorHelpers.test.ts
Comment (seve) [reply client/__tests__/util/stateManager/colorHelpers.test.ts]: It's ?.() not ?(), should work we use it in other places. https://github.com/chanzuckerberg/cellxgene/blob/eaae6df5e34682e5f96aadc3f1d7fc72ba342ca1/client/src/components/continuousLegend/index.js#L140

File: client/src/util/stateManager/colorHelpers.ts
Comment (seve) [reply client/src/util/stateManager/colorHelpers.ts:28]: Since colorMode is passed from reducer state and not the action, shouldn't it be typed as ColorState["colorMode"] ?

File: client/__tests__/util/stateManager/colorHelpers.test.ts
Comment (tihuan) [reply client/__tests__/util/stateManager/colorHelpers.test.ts]: OH! Totally didn't know that's available 😆 Thank you! Updated 👍

File: client/src/util/stateManager/colorHelpers.ts
Comment (tihuan) [reply client/src/util/stateManager/colorHelpers.ts]: Ooh great point! Added `ColorsState` 👌 Thank you!

**Review by seve — APPROVED** [review]

**Comment by tihuan** [comment]
Hi @bkmartinjr ! Wanted to check if you prefer to wait for your review before merging this PR 🙏 Thanks so much!

**Comment by tihuan** [comment]
@maniarathi Actually Bruce is backlogged with other higher pri tasks, are you okay with us merging this PR without his review? Thank you!

**Comment by tihuan** [comment]
I'll merge first and ask Bruce to review later when he's less busy! Thank you!

**Pull request merged**
